### PR TITLE
style(v2): reduce number of ESLint warnings

### PR DESCRIPTION
**Pull request by Josh-Cena**
## Motivation

When I'm working on the project I was constantly overwhelmed (exaggeration here; there aren't so many) by ESLint warnings. Although they aren't serious issues, I decide it's better that we address these issues, since this is what we have the linter for.

There are some warnings that I probably need help with, especially regarding `any` types. (I realize how intractable typing objects can become.) I don't want to easily resort to `eslint-disable` comments.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

These are mostly style changes that don't really impact functionality to any degree.

### CODE REVIEW

File: packages/docusaurus-utils/src/markdownParser.ts [hunk 1]
```diff
@@ -36,17 +36,17 @@ export function createExcerpt(fileString: string): string | undefined {
       // Remove HTML tags.
       .replace(/<[^>]*>/g, '')
       // Remove Title headers
-      .replace(/^\#\s*([^#]*)\s*\#?/gm, '')
+      .replace(/^#\s*([^#]*)\s*#?/gm, '')
```
Comment (Josh-Cena) [inline packages/docusaurus-utils/src/markdownParser.ts:39]: I'm not totally confident with the changes here, but as far as I know, regex is wise enough to know when to treat characters as special ones and when to treat them as plain text. Therefore I just followed the ESLint recommendations.

Comment (slorber) [reply packages/docusaurus-utils/src/markdownParser.ts:39]: agree

File: packages/docusaurus-plugin-content-blog/src/blogUtils.ts [hunk 2]
```diff
@@ -32,7 +32,7 @@ import {LoadContext} from '@docusaurus/types';
 import {validateBlogPostFrontMatter} from './blogFrontMatter';
 
 export function truncate(fileString: string, truncateMarker: RegExp): string {
-  return fileString.split(truncateMarker, 1).shift()!;
+  return fileString.split(truncateMarker, 1).shift() as string;
```
Comment (slorber) [inline packages/docusaurus-plugin-content-blog/src/blogUtils.ts:35]: I don't like this change much. This basically do the exact same thing, so if "as string" does not emit a warning, we'd rather allow usage of !

File: packages/docusaurus-plugin-content-docs/src/sidebars.ts [hunk 3]
```diff
@@ -127,7 +127,10 @@ function assertIsCategory(
     );
   }
   // "collapsed" is an optional property
-  if (item.hasOwnProperty('collapsed') && typeof item.collapsed !== 'boolean') {
+  if (
+    Object.prototype.hasOwnProperty.call(item, 'collapsed') &&
```
Comment (slorber) [inline packages/docusaurus-plugin-content-docs/src/sidebars.ts:131]: too much ceremony, I'd just replace by `typeof item.collapsed !== "undefined"` ?

File: packages/docusaurus-plugin-content-docs/src/versions.ts [hunk 4]
```diff
@@ -288,11 +288,9 @@ function createVersionMetadata({
   // retro-compatible values
   const defaultVersionLabel =
     versionName === CURRENT_VERSION_NAME ? 'Next' : versionName;
-  const defaultVersionPathPart = isLast
-    ? ''
-    : versionName === CURRENT_VERSION_NAME
-    ? 'next'
-    : versionName;
+  const versionNameNotLast =
```
Comment (slorber) [inline packages/docusaurus-plugin-content-docs/src/versions.ts:291]: this name is a bit confusing. What about using a function like getDefaultVersionPathPart instead?

File: packages/docusaurus-theme-bootstrap/src/theme/ThemedImage/index.tsx [hunk 5]
```diff
@@ -21,10 +21,10 @@ const ThemedImage = (props: Props): JSX.Element => {
 
   type SourceName = keyof Props['sources'];
 
+  const clientTheme: SourceName[] = isDarkTheme ? ['dark'] : ['light'];
```
Comment (slorber) [inline packages/docusaurus-theme-bootstrap/src/theme/ThemedImage/index.tsx:24]:
```suggestion
  const clientThemes: SourceName[] = isDarkTheme ? ['dark'] : ['light'];
```

File: packages/docusaurus-theme-classic/src/theme/ThemedImage/index.tsx [hunk 6]
```diff
@@ -21,10 +21,10 @@ const ThemedImage = (props: Props): JSX.Element => {
 
   type SourceName = keyof Props['sources'];
 
+  const clientTheme: SourceName[] = isDarkTheme ? ['dark'] : ['light'];
```
Comment (slorber) [inline packages/docusaurus-theme-classic/src/theme/ThemedImage/index.tsx:24]:
```suggestion
  const clientThemes: SourceName[] = isDarkTheme ? ['dark'] : ['light'];
```

**Review by slorber — CHANGES_REQUESTED** [review]
thanks, not agreeing with everything but overwall happy to reduce eslint warnings

I personally like TS function return type inference but it seems to be a best practice to type return types explicitly so why not

File: packages/docusaurus-plugin-content-blog/src/blogUtils.ts (hunk 2 above)
Comment (Josh-Cena) [reply packages/docusaurus-plugin-content-blog/src/blogUtils.ts:35]: Should I add `eslint-ignore` then...?

**Comment by Josh-Cena** [comment]
@slorber If I'm delving into typing (the `any` types), this PR is likely to shift its focus away from some simple style improvements. Should I do it here or should I open a new PR?

File: packages/docusaurus-plugin-content-docs/src/sidebars.ts (hunk 3 above)
Comment (Josh-Cena) [reply packages/docusaurus-plugin-content-docs/src/sidebars.ts:131]: There is actually a consideration about `collapsed` existing on the prototype chain of `item`. But I think it's safe to assume here that `typeof item.collapsed !== "undefined"` works the same...

File: packages/docusaurus-plugin-content-blog/src/blogUtils.ts (hunk 2 above)
Comment (slorber) [reply packages/docusaurus-plugin-content-blog/src/blogUtils.ts:35]: you can disable the no-non-null rule totally. If we allow "as", we should allow "!", so in any case we'd rather be consistent

File: packages/docusaurus-plugin-content-docs/src/sidebars.ts (hunk 3 above)
Comment (slorber) [reply packages/docusaurus-plugin-content-docs/src/sidebars.ts:131]: yes it looks to me as the sidebar is a regular object

**Comment by slorber** [comment]
> @slorber If I'm delving into typing (the any types), this PR is likely to shift its focus away from some simple style improvements. Should I do it here or should I open a new PR?

smaller PRs are always easier to review/merge, so splitting the work is not a bad idea.

**Comment by Josh-Cena** [comment]
Need some help here...

https://github.com/facebook/docusaurus/blob/7dc9fe839b378bcc570ec56d61295dcb38c09461/packages/docusaurus-plugin-content-docs/src/sidebarItemsGenerator.ts#L270-L274

would this be transformed to a `forEach` loop like this?

```js
docs.forEach(async doc => {await handleDocItem(doc)});
```

Airbnb is complaining about `for...of` loops being too heavy.

**Comment by Josh-Cena** [comment]
I think I'm mostly done, except for the question above... The rest of the warnings (39 of them) contain:

- `Unexpected require()`

https://github.com/facebook/docusaurus/blob/7dc9fe839b378bcc570ec56d61295dcb38c09461/packages/docusaurus-init/templates/classic/src/components/HomepageFeatures.js#L8

This is from the init template and is shipped to the user. The current code looks clean enough and a number of users probably don't have ESLint anyways. How should we safely ignore this issue?

- `iterators/generators require regenerator-runtime, which is too heavyweight for this guide to allow them. Separately, loops should be avoided in favor of array iterations`

As mentioned in the previous comment, I need help to refactor this.

- `Visible, non-interactive elements with click handlers must have at least one keyboard listener`

https://github.com/facebook/docusaurus/blob/7dc9fe839b378bcc570ec56d61295dcb38c09461/packages/docusaurus-theme-classic/src/theme/Toggle/index.tsx#L56-L60

I see that we are planning to add [hotkey binding maps](https://github.com/facebook/docusaurus/issues/3020). This warning will remind us of this feature :P

**Comment by slorber** [comment]
> Need some help here... Airbnb is complaining about for...of loops being too heavy. Would this be transformed to a reduce loop like this? I'm not so confident with async functions.

Honestly don't like Airbnb preset, it's way too opinionated with weird rules, but it's the current docu preset so let's just disable what is annoying. This rule can also be disabled, it's not really applicable for NodeJS code IMHO.


I'll merge this PR as is, any further improvements are welcome

File: packages/docusaurus-plugin-content-docs/src/sidebarItemsGenerator.ts [hunk 7]
```diff
@@ -268,10 +268,10 @@ export const DefaultSidebarItemsGenerator: SidebarItemsGenerator = async functio
     }
 
     // async process made sequential on purpose! order matters
-    for (const doc of docs) {
-      // eslint-disable-next-line no-await-in-loop
-      await handleDocItem(doc);
-    }
+    await docs.reduce(
```
Comment (slorber) [inline packages/docusaurus-plugin-content-docs/src/sidebarItemsGenerator.ts:271]: actually I'd prefer to revert this change

File: packages/docusaurus-plugin-content-docs/src/sidebars.ts (hunk 3 above)
Comment (slorber) [reply packages/docusaurus-plugin-content-docs/src/sidebars.ts:131]: can you replace with `typeof item.collapsed !== "undefined"` ? it has not been done here

File: packages/docusaurus/src/server/configValidation.ts [hunk 8]
```diff
@@ -116,7 +116,9 @@ const SiteUrlSchema = URISchema.required().custom(function (value, helpers) {
         warningMessage: `the url is not supposed to contain a sub-path like '${pathname}', please use the baseUrl field for sub-paths`,
       });
     }
-  } catch (e) {}
+  } catch (e) {
+    /* Do nothing with invalid URL */
```
Comment (slorber) [inline packages/docusaurus/src/server/configValidation.ts:120]: what about using `{ "allowEmptyCatch": true }` instead?

**Review by slorber — CHANGES_REQUESTED** [review]
finally, asking for a few changes 🤪

**Comment by Josh-Cena** [comment]
Okay cool. Lemme address those...

File: packages/docusaurus-plugin-content-docs/src/sidebarItemsGenerator.ts (hunk 7 above)
Comment (Josh-Cena) [reply packages/docusaurus-plugin-content-docs/src/sidebarItemsGenerator.ts]: Although I reverted this, I'm pretty confident that the new code works correctly😂 Just that it is less readable

Comment (slorber) [reply packages/docusaurus-plugin-content-docs/src/sidebarItemsGenerator.ts]: yes definitively works with reduce but much more complicated :)

**Comment by slorber** [comment]
thanks 👍
